exasol_advanced_analytics_framework/interface/create_event_handler_udf.py

```diff
@@ -1,44 +1,93 @@
-from exasol_advanced_analytics_framework.event_handler.event_handler_base import \
-    EventHandlerBase
-from exasol_advanced_analytics_framework.event_handler.event_handler_context import EventHandlerContext
 import importlib
-
-from exasol_advanced_analytics_framework.event_handler.event_handler_result import \
-    EventHandlerResult
+from collections import OrderedDict
+from pathlib import PurePosixPath
+from exasol_bucketfs_utils_python.bucketfs_factory import BucketFSFactory
+from exasol_bucketfs_utils_python.bucketfs_location import BucketFSLocation
+from exasol_advanced_analytics_framework.event_handler.event_handler_base \
+    import EventHandlerBase
+from exasol_advanced_analytics_framework.event_handler.event_handler_context \
+    import EventHandlerContext
+from exasol_advanced_analytics_framework.event_handler.event_handler_result \
+    import EventHandlerResult
+from exasol_advanced_analytics_framework.event_handler.event_handler_state \
+    import EventHandlerState
+from exasol_advanced_analytics_framework.utils.udf_context_wrapper import \
+    UDFContextWrapper
 
 
 class CreateEventHandlerUDF:
+    STATE_DIRECTORY = "aaf_state"
+
     def __init__(self, exa):
         self.exa = exa
 
     def run(self, ctx) -> None:
+        # get and set method parameters
         parameters = ctx.parameters
         bucketfs_connection = ctx.bucketfs_connection
+        bucketfs_location = BucketFSFactory().create_bucketfs_location(
+            url=bucketfs_connection.address,
+            user=bucketfs_connection.user,
+            pwd=bucketfs_connection.password,
+            base_path=parameters["base"]
+        )
+        bucketfs_path = PurePosixPath(
+            self.STATE_DIRECTORY, f"{parameters['bucket_file_path']}.pkl")
+
+        # load the latest (create if not) event handler state object
+        latest_state = self._load_latest_state(
+            parameters, bucketfs_location, bucketfs_path)
+        event_handler_context: EventHandlerContext = latest_state.context
+        event_handler: EventHandlerBase = latest_state.event_handler
 
-        event__handler_context = EventHandlerContext(bucketfs_connection)
-        event_handler_obj = self._get_event_handler_obj(
-            parameters,  event__handler_context)
-        result: EventHandlerResult = event_handler_obj.handle_event()  # TODO ?
-        self._save_event_handler_obj(event_handler_obj, event__handler_context)
+        # call the user code
+        udf_context = self._create_udf_context_wrapper(ctx)
+        result: EventHandlerResult = event_handler.handle_event(
+            udf_context, event_handler_context)  # TODO does user change state?
 
+        # update and save (overwrite)the current state
+        current_state = EventHandlerState(event_handler_context, event_handler)
+        self._save_current_state(
```

Comment:
maybe store new state first to a new file and then remove the old one, such that we have a safe checkpoint

Reply:
since I use same location and path, isn't it overwriting?

Reply:
yes it is overwriting and this is a problem. To keep a safe checkpoint, you need to create a new file and then delete the old one.

Reply:
then we can do something like this that will also solve the following problem:

-  when calling udf query on event-loop side, we can add a parameter like step_num to udf query
- If step_num = 0, we know that there is no saved state yet. So, create new one.
- Adding step_num to the name of state objects makes state objs unique (remove state_obj_[step_num-1], save state_obj_[step_num])

Reply:
Sounds good. For step num 0 we can also add the config as a parameter to the UDF. On that note, I remember something. The UDF for the event handler needs to have dynamic input parameter.